core/src/com/unciv/logic/city/CityConstructions.kt

```diff
@@ -604,10 +640,61 @@ class CityConstructions {
         raisePriority(constructionQueueIndex + 1)
     }
 
-    //endregion
     private fun MutableList<String>.swap(idx1: Int, idx2: Int) {
         val tmp = this[idx1]
         this[idx1] = this[idx2]
         this[idx2] = tmp
     }
+
+    /** Support for [UniqueType.CreatesOneImprovement]:
+     *
+     *  If [building] is an improvement-creating one, find a marked tile matching the improvement to be created
+     *  (skip if none found), then un-mark the tile and place the improvement unless [removeOnly] is set.
+     */
+    fun applyCreateOneImprovement(building: Building, removeOnly: Boolean = false) {
+        val improvement = building.getImprovementToCreate(cityInfo.getRuleset())
+            ?: return
+        val tileForImprovement = getTileForImprovement(improvement.name) ?: return
+        tileForImprovement.stopWorkingOnImprovement()  // clears mark
+        if (removeOnly) return
+        /**todo unify with [UnitActions.getImprovementConstructionActions] and [MapUnit.workOnImprovement] - this won't allow e.g. a building to place a road */
+        tileForImprovement.improvement = improvement.name
+        cityInfo.civInfo.lastSeenImprovement[tileForImprovement.position] = improvement.name
+        cityInfo.cityStats.update()
+        cityInfo.civInfo.updateDetailedCivResources()
+        // If bought the worldscreen will not have been marked to update, and the new improvement won't show until later...
+        if (UncivGame.isCurrentInitialized())
+            UncivGame.Current.worldScreen.shouldUpdate = true
+    }
+
+    /** Support for [UniqueType.CreatesOneImprovement]:
+     *
+     *  To be called after circumstances forced clearing a marker from a tile (pillaging, nuking).
+     *  Should remove one matching building from the queue without looking for a marked tile.
+     */
+    fun removeCreateOneImprovementConstruction(improvement: String) {
+        val ruleset = cityInfo.getRuleset()
+        val indexToRemove = constructionQueue.withIndex().mapNotNull {
+            val construction = getConstruction(it.value)
+            val buildingImprovement = (construction as? Building)?.getImprovementToCreate(ruleset)?.name
+            it.index.takeIf { buildingImprovement == improvement }
+        }.firstOrNull() ?: return
+
+        constructionQueue.removeAt(indexToRemove)
+
+        if (constructionQueue.isEmpty()) {
+            constructionQueue.add("Nothing")
```

Comment:
Const?

Reply:
Yeah, good one, though "not this PR's change" should trump.

Six uses, all the same meaning, so `Constants.kt` _or_ reuse one of them as `PerpetualConstruction.idle.name` the other five places. Not here please.